src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java

```diff
@@ -277,38 +275,111 @@ private static Rule getRepository(
-   * system stays consistent.
-   *
-   * <p>
-   * Returns null if the file system is not up to date and a hash of the marker file if the file
-   * system is up to date.
-   */
-  @Nullable
-  private byte[] isFilesystemUpToDate(Path markerPath, Rule rule, String ruleKey,
-      RepositoryFunction handler, Environment env)
+  private RepositoryDirectoryValue setupOverride(
+      PathFragment sourcePath, Environment env, Path repoRoot)
       throws RepositoryFunctionException, InterruptedException {
-    try {
+    setupRepositoryRoot(repoRoot);
+    RepositoryDirectoryValue.Builder directoryValue = LocalRepositoryFunction.symlink(
+        repoRoot, sourcePath, env);
+    if (directoryValue == null) {
+      return null;
+    }
+    byte[] digest = new byte[]{};
+    return directoryValue.setDigest(digest).build();
+  }
+
+  private static class DigestWriter {
+    private final Path markerPath;
+    private final Rule rule;
+    private final Map<String, String> markerData;
+    private final String ruleKey;
+
+    DigestWriter(BlazeDirectories directories, RepositoryName repositoryName, Rule rule,
+        byte[] ruleSpecificData) {
+      ruleKey = computeRuleKey(rule, ruleSpecificData);
+      markerPath = getMarkerPath(directories, repositoryName.strippedName());
+      this.rule = rule;
+      markerData = Maps.newHashMap();
+    }
+
+    byte[] writeMarkerFile() throws RepositoryFunctionException {
+      StringBuilder builder = new StringBuilder();
+      builder.append(ruleKey).append("\n");
+      for (Map.Entry<String, String> data : markerData.entrySet()) {
+        String key = data.getKey();
+        String value = data.getValue();
+        builder.append(escape(key)).append(" ").append(escape(value)).append("\n");
+      }
+      String content = builder.toString();
+      try {
+        FileSystemUtils.writeContent(markerPath, StandardCharsets.UTF_8, content);
+      } catch (IOException e) {
+        throw new RepositoryFunctionException(e, Transience.TRANSIENT);
+      }
+      return new Fingerprint().addString(content).digestAndReset();
+    }
+
+    /**
+     * Checks if the state of the repository in the file system is consistent with the rule in the
+     * WORKSPACE file.
+     *
+     * <p>
+     * Deletes the marker file if not so that no matter what happens after, the state of the file
+     * system stays consistent.
+     *
+     * <p>
+     * Returns null if the file system is not up to date and a hash of the marker file if the file
+     * system is up to date.
+     */
+    byte[] isFileSystemUpToDate(RepositoryFunction handler, Environment env)
+        throws RepositoryFunctionException, InterruptedException {
       if (!markerPath.exists()) {
         return null;
       }
 
-      String content = FileSystemUtils.readContent(markerPath, StandardCharsets.UTF_8);
+      Map<String, String> markerData = new TreeMap<>();
+      String content;
+      try {
```

Comment:
nit: why not wrap the whole thing into one big try-catch block instead of having three of those?

Reply:
Done